src/voter.rs

```diff
@@ -441,16 +465,203 @@ impl<H, N, E: Environment<H, N>> Future for BackgroundRound<H, N, E> where
 
 	fn poll(&mut self) -> Poll<u64, E::Error> {
 		self.task = Some(::futures::task::current());
-		self.inner.poll()?;
 
-		if self.is_done() {
-			Ok(Async::Ready(self.inner.votes.number()))
+		let mut voting_round = self.inner.lock();
+		voting_round.poll()?;
+
+		if self.is_done(&voting_round) {
+			Ok(Async::Ready(voting_round.votes.number()))
 		} else {
 			Ok(Async::NotReady)
 		}
 	}
 }
 
+pub struct RoundCommitter<H, N, E: Environment<H, N>> where
+	H: Hash + Clone + Eq + Ord + ::std::fmt::Debug,
+	N: Copy + BlockNumberOps + ::std::fmt::Debug,
+{
+	voting_round: Arc<Mutex<VotingRound<H, N, E>>>,
+	commit_timer: E::Timer,
+	last_commit: Option<Commit<H, N, E::Signature, E::Id>>,
+}
+
+impl<H, N, E: Environment<H, N>> RoundCommitter<H, N, E> where
+	H: Hash + Clone + Eq + Ord + ::std::fmt::Debug,
+	N: Copy + BlockNumberOps + ::std::fmt::Debug,
+{
+	fn import_commit(
+		&mut self,
+		env: &E,
+		commit: Commit<H, N, E::Signature, E::Id>,
+	) -> Result<bool, E::Error> {
+		let mut voting_round = self.voting_round.lock();
+
+		// ignore commits for a lower block than we already finalized
+		if commit.target_number < voting_round.votes.finalized().map(|(_, n)| *n).unwrap_or(N::zero()) {
+			return Ok(true);
+		}
+
+		// check that all precommits are for blocks higher than the target
+		// commit block, and that they're its descendents
+		if !commit.justification.iter().all(|signed| {
+			signed.precommit.target_number >= commit.target_number &&
+				env.ancestry(
```

Comment:
precommits could also be for the block itself, in which case it would be `target_hash == commit.target_hash || ancestry(...).is_ok()`